### ISSUE TITLE
Reading : DyLoRA. Parameter-Efficient Tuning of Pretrained Models using Dynamic Search-Free Low Rank Adaptation

### ISSUE THREAD

**Issue by tikiti-kyotaro** [tikiti-kyotaro/Paperlog#13]
**Abstract**
With the ever-growing size of pretrained mod- els (PMs), fine-tuning them has become more expensive and resource-hungry. As a remedy, low-rank adapters (LoRA) keep the main pre- trained weights of the model frozen and just introduce some learnable truncated SVD modules (so-called LoRA blocks) to the model. While LoRA blocks are parameter-efficient, they suffer from two major problems: first, the size of these blocks is fixed and cannot be modified after training (for example, if we need to change the rank of LoRA blocks, then we need to re- train them from scratch); second, optimizing their rank requires an exhaustive search and effort.  In this work, we introduce a dynamic low-rank adaptation (DyLoRA) technique to address these two problems together. Our Dy- LoRA method trains LoRA blocks for a range of ranks instead of a single rank by sorting the representation learned by the adapter module at different ranks during training. We evaluate our solution on different natural language under- standing (GLUE benchmark) and language generation tasks (E2E, DART and WebNLG) using different pretrained models such as RoBERTa and GPT with different sizes. Our results show that we can train dynamic search-free models with DyLoRA at least 4 to 7 times faster than LoRA without significantly compromising performance. Moreover, our models can perform consistently well on a much larger range of ranks compared to LoRA. 

# 1. どんなもの？
LoRAの問題点
1. ランクを変更する場合再訓練する必要がある
2. ランクを最適化する際に網羅的探索と労力が必要となる

DyLoRAを提案
固定のランクに対しての学習ではなくランクの範囲に対して学習を行う

# 2. 先行研究と比べてどこがすごい？
LoRAから特にコストを大幅に増やすことなく動的にする
ハイパラ探索が簡単になる
LoRAは訓練された特定のランクでしか動かない→しかしランクによってだいぶ性能が変わってしまう

LoRAは低ランクの時は性能が下がってしまうが、DyLoRAなら低ランクの時も性能がそこまで下がらない

Table5 通常のLoRAなら64個のうち全てに対して性能を測らなければ一番いいランクを見つけることができない。今回示す表ではLoRAは1,2,4,8,16,32,64の7回の探索を行なっているが、DyLoRAは1回の探索で同程度の性能を誇る

Table2 DyLoRAはLoRAに比べてロバスト性が見られる

# 3.技術や手法のキモはどこ？
学習中に異なるランクで学習した表現を整理することで、単一のランクではなく、様々なランクに対してLoRAブロックを訓練する
(Figure1)

Nested Dropout
通常のDropoutでは順番を考慮しないが、Nested Dropoutでは順番を考慮する
→通常のDropoutは割合を決めたらランダムに不活性化していくが、Nested Dropoutは
→学習時に重要な情報を低いインデックスに入れる

特定のランクではなくランクの範囲に対して訓練する(r_minとr_maxを新たなハイパラとしてその間の範囲を訓練する)→ランクを変えても性能はそこまで変わらないことを確認
→LoRAの学習過程で異なるランクの情報の内容をソートすることが考えられる

**eq5** それぞれの重みに対してrをbで区切ってサンプリングする (fig1の真ん中の図)
**eq6** それぞれのb番目（境界線の部分）の値を取っておく
**eq7** eq5で区切った重み行列でLoRAを行う

(Table6)

# 4. どうやって有効だと検証した？
タスク：NLU, NLG
モデル：RoBERTa, GPT2-medium
ベースライン：FT, LoRA, FLOP

Table3  ??
Table6 生成タスクでも同様にLoRAよりも高い性能かつ広いランクを誇った

アブレーション
確率分布を変えてみる
geomeric は低ランクの時にいい性能をだし、uniformは全てのランクでいい性能を持つ

# 5. 議論はある？

# 6.次に読むべき研究は？
